### ISSUE TITLE
./waf raises RuntimeError

### ISSUE THREAD

**Issue by dfyockey** [kupferlauncher/kupfer#125]
Following successful completion of the `./waf configure` installation step, execution of the build step using either `./waf` or `python3 ./waf` raises the error `RuntimeError: generator raised StopIteration`.

### Environment

Kupfer Version: v319
Window Manager: Xfwm4
Desktop Environment: Xfce
Linux Distribution: Fedora 31
Python Version: 3.7.6

### Actual Behaviour
```
Traceback (most recent call last):
  File "/home/David/Downloads/kupfer_test/kupfer-v319/waflib/Node.py", line 490, in ant_iter
    raise StopIteration
StopIteration

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/David/Downloads/kupfer_test/kupfer-v319/waflib/Scripting.py", line 138, in waf_entry_point
    run_commands()
  File "/home/David/Downloads/kupfer_test/kupfer-v319/waflib/Scripting.py", line 231, in run_commands
    run_command(cmd_name)
  File "/home/David/Downloads/kupfer_test/kupfer-v319/waflib/Scripting.py", line 216, in run_command
    ctx.execute()
  File "/home/David/Downloads/kupfer_test/kupfer-v319/waflib/Scripting.py", line 551, in execute
    return execute_method(self)
  File "/home/David/Downloads/kupfer_test/kupfer-v319/waflib/Build.py", line 218, in execute
    self.load_envs()
  File "/home/David/Downloads/kupfer_test/kupfer-v319/waflib/Build.py", line 179, in load_envs
    lst = node.ant_glob('**/*%s' % CACHE_SUFFIX, quiet=True)
  File "/home/David/Downloads/kupfer_test/kupfer-v319/waflib/Node.py", line 578, in ant_glob
    ret = [x for x in self.ant_iter(accept=accept, pats=[to_pat(incl), to_pat(excl)], maxdepth=25, dir=dir, src=src, remove=kw.get('remove', True))]
  File "/home/David/Downloads/kupfer_test/kupfer-v319/waflib/Node.py", line 578, in <listcomp>
    ret = [x for x in self.ant_iter(accept=accept, pats=[to_pat(incl), to_pat(excl)], maxdepth=25, dir=dir, src=src, remove=kw.get('remove', True))]
RuntimeError: generator raised StopIteration
```

### Apparent Cause
[PEP 479 -- Change StopIteration handling inside generators](https://www.python.org/dev/peps/pep-0479). According to PEP 479, when StopIteration is raised inside a generator, it is replaced with RuntimeError. As understood from the PEP 479 [Transition plan](https://www.python.org/dev/peps/pep-0479/#transition-plan) and from [What’s New In Python 3.7 : Porting to Python 3.7 : Changes in Python Behavior](https://docs.python.org/3/whatsnew/3.7.html#changes-in-python-behavior), deprecated behavior preventing the RuntimeError was removed in Python 3.7.

### Workaround
Replace `raise StopIteration` on line 490 of `waflib/Node.py` with `return`, as indicated at PEP 479 [Consequences for existing code : Writing backwards and forwards compatible code](https://www.python.org/dev/peps/pep-0479/#writing-backwards-and-forwards-compatible-code).

### Expected Behaviour (and result upon Workaround implementation)
Successful complition of build.

**Comment by KarolBedkowski** [comment]
Yes, WAF used in Kupfer is quite old.
Updated version for test: https://github.com/KarolBedkowski/kupfer/tree/waf-update

**Comment by dfyockey** [comment]
Tested the updated version, and it unfortunately gave an error... 
```
[David@Fedora64 kupfer-waf-update]$ ./waf
Waf: Entering directory `/home/David/Downloads/kupfer-waf-update/build'
Waf: Leaving directory `/home/David/Downloads/kupfer-waf-update/build'
Redundant -c flag in intltool task bld(source='kupfer.desktop.in', target='kupfer.desktop', meths=['apply_intltool_in_f', 'process_rule'], features=['intltool_in'], path=/home/David/Downloads/kupfer-waf-update/auxdata, idx=1, tg_idx_count=40, podir='../po', flags=('-d', '-q', '-u', '-c'), install_path='${DATADIR}/applications', chmod=493, posted=True) in /home/David/Downloads/kupfer-waf-update/auxdata
[David@Fedora64 kupfer-waf-update]$
```
The same resulted with `python3 waf`.

Throwing caution to the wind, ignoring the error, and attempting to install anyway resulted in the same error...
```
[David@Fedora64 kupfer-waf-update]$ sudo ./waf install
[sudo] password for David: 
Waf: Entering directory `/home/David/Downloads/kupfer-waf-update/build'
Waf: Leaving directory `/home/David/Downloads/kupfer-waf-update/build'
Redundant -c flag in intltool task bld(source='kupfer.desktop.in', target='kupfer.desktop', meths=['apply_intltool_in_f', 'process_rule'], features=['intltool_in'], path=/home/David/Downloads/kupfer-waf-update/auxdata, idx=1, tg_idx_count=40, podir='../po', flags=('-d', '-q', '-u', '-c'), install_path='${DATADIR}/applications', chmod=493, posted=True) in /home/David/Downloads/kupfer-waf-update/auxdata
[David@Fedora64 kupfer-waf-update]$
```
...and, of course, the application wasn't installed.

**Comment by KarolBedkowski** [comment]
Works for me in Python 3.7.
Try run `python3 waf clean` first.
"Redundant -c flag..." error can be ignored I hope).

**Comment by dfyockey** [comment]
Tried `python3 waf clean`, but it made no difference.

Considering the "Redundant -c flag..." error, I searched the contents of the `kupfer-waf-update` files and located the flags in `kupfer-waf-update/auxdata/wscript`. The build succeeded after deleting the `-c` flag from each of lines 34 and 59 of the file; the line 59 deletion was prompted by a similar error that arose after the line 34 deletion.

Installation succeeded after the successful build, but the application would not run. It instead generated the following error...
```
[David@Fedora64 kupfer-waf-update]$ kupfer
Traceback (most recent call last):
  File "/usr/local/share/kupfer/kupfer.py", line 21, in <module>
    from kupfer import main
ImportError: cannot import name 'main' from 'kupfer' (/usr/local/share/kupfer/kupfer.py)
[David@Fedora64 kupfer-waf-update]$ 

```
Any chance this could somehow be due to the deletion of the `-c` flags?

**Comment by KarolBedkowski** [comment]
Yes, auxdata should be also updated. 
I think now work:
https://github.com/KarolBedkowski/kupfer/tree/waf-update

**Comment by dfyockey** [comment]
Tried it, but it wouldn't install at all until installing `python3-docutils` to provide `rst2man`, which `./waf configure` results had said was missing but optional.

Even though the installation _seemed_ to complete, there was no statement that it was successful. Instead, the following error was generated:
```
source not found: 'generalusage.page' in bld(source=[], target='', meths=['process_install_task', 'process_rule', 'process_source'], features=['install_task'], path=/home/David/Downloads/kupfer-waf-update, idx=169, tg_idx_count=178, install_to='/usr/local/share/help/C/kupfer', install_from='generalusage.page', dest='/usr/local/share/help/C/kupfer', type='install_files', posted=True, install_task=
	{task 140152911199696: inst  -> }) in /home/David/Downloads/kupfer-waf-update
```
A substantially identical error had also been generated by the failed install prior to providing `rst2man`.  References to `generalusage.page` and `help` in the error had suggested to me the need for `rst2man`, but the error was generated even after it was provided.

Trying to run kupfer after installation resulted in the same Trackback and ImportError as before. However, I _was_ able to get the application running by examining the contents of the installed folders and moving some of the files in the installation. It seems that the install script is erroneously creating subdirectories in the directories where files should be placed and then erroneously placing those files in the subdirectories. A few of examples:
- Files that should be in `/usr/local/share/kupfer/kupfer` are installed in `/usr/local/share/kupfer/kupfer/kupfer`.
- Files that should be in `/usr/local/share/kupfer/kupfer/core` are installed in `/usr/local/share/kupfer/kupfer/core/kupfer/core`
- Files that should be in `/usr/local/share/kupfer/kupfer/plugin/core` are installed in `/usr/local/share/kupfer/kupfer/plugin/core/kupfer/plugin/core`

My thought is that if the erroneous directory creation and erroneous file placement were corrected, all may be good.

Incidentally, all directories created during installation seem to have 744 permissions, making it necessary to run the application as root. It can be worked around fairly easily by running an appropriate `sudo chmod` command to change them all to, e.g., 755, but are the 744 permissions intentional? Just wondering.

**Comment by KarolBedkowski** [comment]
Ok,
I hope last update solve all this problem except permissions.
Waf install copy files so the should have permission rX for group/others. Also umask should be set to i.e. 0022. I don't see any standard way to change it.

**Comment by dfyockey** [comment]
Sorry about the delayed response; was out of town for several days.

Configuration, build and install were all successful, and the main application seems to run fine. So, we're making progress! :^)

But some files are still falling into incorrect locations. Perhaps most problematic, the file `defaults.cfg` and the `.ui` files that should all be in `/usr/local/share/kupfer` were installed one directory up in `/usr/local/share`; consequently, the Preferences dialog isn't available (and presumably other features dependant on those files are unavailable as well).

I also noticed that the image files which should be in `/usr/local/share/kupfer/kupfer/plugin/core` and `/usr/local/share/kupfer/kupfer/plugin/qsicons` are all instead in `/usr/local/share/kupfer`. Other files may be mislocated as well.

I imagine that this must be getting frustrating, so I've got to say... Thanks for all the effort you're putting in! I really appreciate it!
